source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,70 @@
+
+## Changes to Engine and Vdsm
+
+There are no special changes needed in order to add a TPM device as such, it's just another kind of device added to Engine.
+
+But a very different thing is ensuring TPM data persistence.  TPM data is managed using an external tool, swtpm.  Once the VM domain XML is deleted on the host, the TPM data is deleted as well by libvirt and it must be restored before the VM is run again, whether on this or another host (live host migrations are handled transparently and don't require any special actions on the oVirt side).
+
+In order to support TPM data persistence, we must store the data in the Engine database and transfer it to the host before the VM is started and update it in the Engine database after the VM gets down.  We need the following Vdsm API changes to achieve that:
+
+- Adding a new VM.create parameter providing the data.
+- Adding a new parameter to VM.destroy call response containing the updated data.
+
+There is some fragility with that approach.  If VM.destroy call fails for any reason after the libvirt domain is undefined (and the corresponding TPM data is deleted by libvirt), the updated data will be lost.  But we don't have many alternatives: While we could make a separate API call to retrieve TPM data or attach them to VM stats, that would mean retrieving live data from swtpm, which may not be consistent.  Only data retrieved after the VM stops can be considered stable.  And even with a separate API call, if Engine loses contact with the host and starts the VM elsewhere, updates to TPM data since the VM start will be lost.  We cannot guarantee, without introducing complicated implementations, up-to-date TPM data if the environment doesn't work as expected.  Considering TPM data typically serves forensic purposes rather than storing critical information, we may accept the fact that VM guest OS repair is needed in case the data got changed and the environment experienced unexpected trouble (i.e. VM.destroy call hasn't been successful).  It may still cause trouble for HA VMs.
+
+An alternative mechanism could be to use shared storage to store TPM data, either by copying the data there or (preferably, if possible) to let swtpm use the shared storage directly.  The differences to the solution proposed above would be:
+
+- More complicated implementation.
+- Need to connect stored data to the corresponding (disk-only) snapshots, either on the storage, or by copying them to and from the database.
+- Better data safety in case of direct use of the shared storage by swtpm, the data is updated immediately, not only once VM stops running (but there may in theory data consistency problems again).
+- It's unclear how swtpm would work with network storage and how to prevent libvirt from deleting the data once the domain XML is deleted on the host.
+
+Overall, storing TPM data in the database looks much easier and so also more robust in the end result.
```

Comment:
Can you please add also a section on how to test the feature?

Reply:
Done.